bittensor/core/extrinsics/utils.py

```diff
@@ -50,55 +45,53 @@ def submit_extrinsic(
     extrinsic_hash = extrinsic.extrinsic_hash
     starting_block = substrate.get_block()
 
-    def _handler():
-        """
-        Timeout handler for threading. Will raise a TimeoutError if timeout is exceeded.
-        """
-        logging.error("Timed out waiting for extrinsic submission.")
-        raise _ThreadingTimeoutException
-
-    # sets a timeout timer for the next call to 200 seconds
-    # will raise a _ThreadingTimeoutException if it reaches this point
-    timer = threading.Timer(200, _handler)
-
-    try:
-        timer.start()
-        response = substrate.submit_extrinsic(
-            extrinsic,
-            wait_for_inclusion=wait_for_inclusion,
-            wait_for_finalization=wait_for_finalization,
-        )
-    except SubstrateRequestException as e:
-        logging.error(format_error_message(e.args[0], substrate=substrate))
-        # Re-rise the exception for retrying of the extrinsic call. If we remove the retry logic, the raise will need
-        # to be removed.
-        raise
-
-    except _ThreadingTimeoutException:
-        after_timeout_block = substrate.get_block()
-
+    timeout = EXTRINSIC_SUBMISSION_TIMEOUT
+    event = threading.Event()
+
+    def submit():
+        try:
+            response_ = substrate.submit_extrinsic(
+                extrinsic,
+                wait_for_inclusion=wait_for_inclusion,
+                wait_for_finalization=wait_for_finalization,
+            )
+        except SubstrateRequestException as e:
+            logging.error(format_error_message(e.args[0], substrate=substrate))
+            # Re-raise the exception for retrying of the extrinsic call. If we remove the retry logic,
+            # the raise will need to be removed.
+            raise
+        finally:
+            event.set()
+        return response_
+
+    with ThreadPoolExecutor(max_workers=1) as executor:
         response = None
-        for block_num in range(
-            starting_block["header"]["number"],
-            after_timeout_block["header"]["number"] + 1,
-        ):
-            block_hash = substrate.get_block_hash(block_num)
-            try:
-                response = substrate.retrieve_extrinsic_by_hash(
-                    block_hash, f"0x{extrinsic_hash.hex()}"
+        future = executor.submit(submit)
+        if not event.wait(timeout):
+            logging.error("Timed out waiting for extrinsic submission.")
+            after_timeout_block = substrate.get_block()
+
+            for block_num in range(
```

Comment:
Not a blocker, but we should rethink this method of getting the block hash.

Reply:
I'm open to suggestions. Why is that?